cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/schedule/store/DatasetBasedStreamSizeScheduleStore.java

```diff
@@ -350,25 +385,98 @@ public void apply() throws Exception {
+        LOG.trace(errorMessage, e);
+      } else {
+        LOG.error(errorMessage, e);
+      }
+    }
+
+    // If upgrade is already complete, then simply return.
+    if (upgradeComplete.get()) {
+      return;
+    }
+
+    final AtomicInteger maxRows = new AtomicInteger(10);
+    final AtomicInteger sleepTimeInMins = new AtomicInteger(1);
+    LOG.info("Starting upgrade of StreamSizeScheduleStore.");
+    while (!isUpgradeComplete()) {
+      sleepTimeInMins.set(1);
+      try {
+        factory.createExecutor(ImmutableList.of((TransactionAware) table))
+          .execute(new TransactionExecutor.Subroutine() {
+            @Override
+            public void apply() {
+              if (upgradeVersionKeys(table, maxRows.get())) {
+                // Upgrade is complete. Mark that app version upgrade is complete in the table.
+                table.put(APP_VERSION_UPGRADE_KEY, Bytes.toBytes('c'),
+                          Bytes.toBytes(ProjectInfo.getVersion().toString()));
+              }
+            }
+          });
+      } catch (TransactionFailureException e) {
+        if (e instanceof TransactionConflictException) {
+          LOG.debug("Upgrade step faced Transaction Conflict exception. Retrying operation now.", e);
+          sleepTimeInMins.set(0);
+        } else if (e instanceof TransactionNotInProgressException) {
+          int currMaxRows = maxRows.get();
+          if (currMaxRows > 1) {
+            maxRows.decrementAndGet();
+          }
+          sleepTimeInMins.set(0);
+          LOG.debug("Upgrade step faced a Transaction Timeout exception. " +
+                      "Reducing the number of max rows to : {} and retrying the operation now.", maxRows.get(), e);
+        } else {
+          LOG.error("Upgrade step faced exception. Will retry operation after some delay.", e);
+          sleepTimeInMins.set(1);
         }
-      });
+      }
+      TimeUnit.MINUTES.sleep(sleepTimeInMins.get());
+    }
+    LOG.info("Upgrade of StreamSizeScheduleStore is complete.");
+  }
+
+  // Returns true if the upgrade flag is set. Upgrade could have completed earlier than this since this flag is
+  // updated asynchronously.
+  public boolean isUpgradeComplete() {
+    return appVerUpgrade.get();
   }
 
-  private void upgradeVersionKeys() {
+  // Return whether the upgrade process is complete - determined by checking if there were no rows that were
+  // upgraded after the invocation of this method.
+  private boolean upgradeVersionKeys(Table table, int maxNumRows) {
+    int numRowsUpgraded = 0;
     Joiner joiner = Joiner.on(":");
     try (Scanner scan = getScannerWithPrefix(KEY_PREFIX)) {
       Row next;
-      while ((next = scan.next()) != null) {
+      // Upgrade only 10 rows in one transaction to reduce the probability of conflicts with regular Store operations.
+      while (((next = scan.next()) != null) && (numRowsUpgraded < maxNumRows)) {
         if (isInvalidRow(next)) {
```

Comment:
If all the rows are invalid for this batch then `numRowsUpgraded` will be zero even though there are records to be upgraded, right?